File: python/fastapi-sqlalchemy/schemas/adoption.py
from sqlalchemy import Column, ForeignKey, Integer, String, Date
from sqlalchemy.orm import relationship
from db.db import Base
from schemas.pet import Pet
from schemas.adopter import Adopter
from schemas.user import User


class Adoption(Base):
    __tablename__ = "adopciones"

    id_adoption = Column(
        name="codigoadopcion",
        type_=Integer,
        autoincrement="auto",
        primary_key=True,
        nullable=False,
    )
    id_pet = Column(Integer, ForeignKey("mascotas.codigomascota"), name="codigomascota")
    id_adopter = Column(
        Integer,
        ForeignKey("adoptadores.codigoadoptador"),
        name="codigoadoptador",
        nullable=False,
    )
    date_adoption = Column(name="fechaadopcion", type_=Date)
    id_user = Column(
        Integer, ForeignKey("usuarios.codigousuario"), name="codigousuario"
    )
    description = Column(name="observaciones", type_=String(length=256))

    pets = relationship("Pet", back_populates="adoptions")
    adopters = relationship("Adopter", back_populates="adoptions")
    users = relationship("User", back_populates="adoptions")

File: javascript/express-pet-shelter/models/entities/adoption.js
import { DataTypes } from "sequelize";
import { sequelize } from "../../db/db.js";
import { Pet } from "./pet.js";
import { Adopter } from "./adopter.js";
import { User } from "./user.js";

export const Adoption = sequelize.define(
  "adopciones",
  {
    adoptionId: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
      field: "codigoadopcion",
    },
    petId: {
      type: DataTypes.INTEGER,
      field: "codigomascota",
    },
    adopterId: {
      type: DataTypes.INTEGER,
      field: "codigoadoptador",
    },
    adoptionDate: {
      type: DataTypes.DATEONLY,
      field: "fechaadopcion",
    },
    userId: {
      type: DataTypes.INTEGER,
      field: "codigousuario",
    },
    observations: {
      type: DataTypes.STRING,
      field: "observaciones",
    },
  },
  {
    timestamps: false,
  }
);

Adoption.belongsTo(Pet, {
  foreignKey: "petId",
  targetKey: "petId",
});
Pet.hasOne(Adoption, {
  foreignKey: "petId",
  sourceKey: "petId",
});

Adoption.belongsTo(User, {
  foreignKey: "userId",
  targetKey: "userId",
});
User.hasOne(Adoption, {
  foreignKey: "userId",
  sourceKey: "userId",
});

Adoption.belongsTo(Adopter, {
  foreignKey: "adopterId",
  targetKey: "adopterId",
});
Adopter.hasOne(Adoption, {
  foreignKey: "adopterId",
  sourceKey: "adopterId",
});
